### ISSUE TITLE
Semi-blast: Index of a protein should not be derived from the NFTID

### ISSUE THREAD

**Issue by zenodeapp** [zenodeapp/protein-crud#25]
In the semi-blast algorithm I've quite often used the logic `index == nftId - 1`. This works fine if we're working with the first 10k (or whatever size) NFT IDs, but is not going to work if NFTs get bundled together in contracts that are not the first batch of NFTs. And will especially not work if they're **not gapless** (e.g. NFT IDs 7 _gap_ 56 _gap_ 300, 301, 302 _gap_ 305, etc.).

Therefore something else should be thought of.

Luckily this problem is only in the Semi-blast algorithm contract.

**Comment by zenodeapp** [comment]
Two possible ways of fixing this:
- We either let the algorithms return indices instead of NFT ID's. The problem that arises when taking this approach is the rise in the amount of external calls that will have to be made in semi-blast (for each matching NFT it would have to look up its index).
- The other approach is to keep track of the highest NFT ID added (to get a ceiling for our memory arrays). This would create unnecessary huge arrays if the NFT IDs are large.

**Comment by zenodeapp** [comment]
UPDATE: I feel like I need to explain this a bit better. In the algorithms I use arrays to keep track of which NFTs match the query. This boosts the speed of the algorithm, for it then skips those we've already processed. The indices for the array equal `nftId - 1` and hold false/true values. So if position 10 in the array is set to true, this would mean that NFT ID 11 was a successful match. This gives us O(1) time when checking whether an NFT was already processed or not.

Now, imagine adding NFTIDs 100-200 in the contract. I initialize an array with proteinCount (which in this case is 100). Querying in this dataset would result in an error for it would check for NFTIDs outside the boundaries of the array (0-99).

**Comment by zenodeapp** [comment]
Two possible ways of fixing this:
- We either let the algorithms return indices instead of NFT ID's. The problem that arises when taking this approach is the rise in the amount of external calls that will have to be made in semi-blast (for each matching NFT it would have to look up its index in the protein indexer contract).
- The other approach is to keep track of the highest NFT ID added (to get a ceiling for our memory arrays). Though, this would create unnecessary huge arrays if the NFT IDs are large.

**Comment by zenodeapp** [comment]
I went with the second approach. Initializing large arrays will take more time, but will probably take less time than to make lots of external calls. This actually would have to be tested.

This does imply that a contract containing a protein with a large nft id will slow down the query functionality.

For this reason and to make sure we're future-proof, I've added functions that would allow the developer to implement the first approach (see above), without needing to change and redeploy the protein indexer contract.
